src/Trie.mo

```diff
@@ -1,13 +1,78 @@
-///
-/// ## Background
+/// independent of operation history (unlike other popular search trees).
 ///
 /// The representation we use here comes from Section 6 of ["Incremental computation via function caching", Pugh & Teitelbaum](https://dl.acm.org/citation.cfm?id=75305).
 ///
+/// ## User's overview
+///
+/// This module provides an applicative (functional) hash map.
+/// Notably, each `put` produces a **new trie _and value being replaced, if any_**.
+///
+/// Those looking for more familar (imperative,
+/// object-oriented) hash map should consider `TrieMap` or `HashMap` instead.
+///
+/// The basic `Trie` operations consist of:
+/// - `put` - put a key-value into the trie, producing a new version.
+/// - `get` - get a key's value from the trie, or `null` if none.
+/// - `iter` - visit every key-value in the trie.
+///
+/// The `put` and `get` operations work over `Key` records,
+/// which group the hash of the key with its non-hash key value.
+///
+/// ```motoko
+/// import Trie "mo:base/Trie";
+/// import Text "mo:base/Text";
+///
+/// type Trie<K, V> = Trie.Trie<K, V>;
+/// type Key<K> = Trie.Key<K>;
+///
+/// func key(t: Text) : Key<Text> { { key = t; hash = Text.hash t } };
+///
+/// let t0 : Trie<Text, Nat> = Trie.empty();
+/// let t1 : Trie<Text, Nat> = Trie.put(t0, key "hello", Text.equal, 42).0;
+/// let t2 : Trie<Text, Nat> = Trie.put(t1, key "world", Text.equal, 24).0;
+/// let n : ?Nat = Trie.put(t1, key "hello", Text.equal, 0).1;
+/// assert (n == ?42);
+/// ```
+///
+/// ## Implementation overview
+///
+/// A (hash) trie is a binary tree container for key-value pairs that
+/// consists of leaf and branch nodes.
+///
+/// Each internal **branch node**
+/// represents having distinguished its key-value pairs on a single bit of
+/// the keys.
+/// By following paths in the trie, we determine an increasingly smaller
+/// and smaller subset of the keys.
+///
+/// Each **leaf node** consists of an association list of key-value pairs.
+///
+/// Each non-empty trie node stores a size; we discuss that more below.
+///
+/// ### Adaptive depth
+///
+/// We say that a leaf is valid if it contains no more than `MAX_LEAF_SIZE`
+/// key-value pairs.  When a leaf node grows too large, the
+/// binary tree produces a new internal binary node, and splits the leaf into
+/// a pair of leaves using an additional bit of their keys' hash strings.
+///
+/// For small mappings, the trie structure consists of a single
+/// leaf, which contains up to MAX_LEAF_SIZE key-value pairs.
+///
+/// ### Cached sizes
 ///
+/// At each branch and leaf, we use a stored size to support a
+/// memory-efficient `toArray` function, which itself relies on
+/// per-element projection via `nth`; in turn, `nth` directly uses the
+/// O(1)-time function `size` for achieving an acceptable level of
+/// algorithmic efficiently.  Notably, leaves are generally lists of
+/// key-value pairs, and we do not store a size for each Cons cell in the
+/// list.
```

Comment:
If you put the implementation notes in simple comments, they might not appear in the doc (is that too much detail for users who don't care about the implementation?)

Reply:
changed this all to `//` as per the suggestion.